syft/messaging/plan.py

```diff
@@ -74,210 +65,326 @@ class State(object):
-        if isinstance(key, (list, tuple)):
-            for k in key:
-                self.append(k)
-        elif not isinstance(key, str):
-            raise ValueError(
-                "Don't provide the element to the state but just its name:\n"
-                "Example: if you have self.elem1, call self.add_to_state('elem1')."
-            )
-        else:
-            t = getattr(self.plan, key)
-            if self.is_valid_type(t):
-                self.keys.add(key)
-            else:
-                raise ValueError(
-                    f"Obj of type {type(t)} is not supported in the state.\n"
-                    "Use instead tensors or parameters."
-                )
-        return self
+        Return state elements
+        """
+        tensors = []
+        for state_id in self.state_ids:
+            tensor = self.owner.get_obj(state_id)
+            tensors.append(tensor)
+        return tensors
 
-    def __iadd__(self, other_keys):
-        return self.append(other_keys)
+    def set_(self, state_dict):
+        """
+        Reset inplace the state by feeding it a dict of tensors or params
+        """
+        assert list(self.state_ids) == list(state_dict.keys())
 
-    def is_valid_type(self, obj):
-        return isinstance(obj, (torch.nn.Module, torch.Tensor))
+        for state_id, new_tensor in state_dict.items():
+            tensor = self.owner.get_obj(state_id)
 
-    def get_id_at_location(self):
+            with torch.no_grad():
+                tensor.set_(new_tensor)
+
+            tensor.child = new_tensor.child if new_tensor.is_wrapper else None
+            tensor.is_wrapper = new_tensor.is_wrapper
+            if tensor.child is None:
+                delattr(tensor, "child")
+
+    @staticmethod
+    def create_grad_if_missing(tensor):
+        if isinstance(tensor, torch.nn.Parameter) and tensor.grad is None:
+            o = tensor.sum()
+            o.backward()
+            if tensor.grad is not None:
+                tensor.grad -= tensor.grad
+
+    def send_for_build(self, location, **kwargs):
         """
-        Returns all the id_at_location of the pointers in the state
+        Send functionality that can only be used when sending the state for
+        building the plan. Other than this, you shouldn't need to send the
+        state separately.
         """
-        id_at_location = []
+        assert location.id == self.plan.id  # ensure this is a send for the build
 
-        for key in self.keys:
-            ptr = getattr(self.plan, key)
-            if isinstance(ptr, torch.nn.Module):
-                for param in ptr.parameters():
-                    id_at_location.append(param.id_at_location)
-            else:
-                id_at_location.append(ptr.id_at_location)
+        for tensor in self.tensors():
+            self.create_grad_if_missing(tensor)
+            tensor.send_(location, **kwargs)
 
-        return id_at_location
+    def fix_precision_(self, *args, **kwargs):
```

Comment:
I'd rather not to: until we have a reliable code I would like the user to manually define the appropriate transformations (like send in inputs to the right worker, transform to fixed prec of additive sharing, etc)
Otherwise some bugs might be hard to find!